algorithmic_efficiency/workloads/wmt/wmt_pytorch/models.py

```diff
@@ -209,6 +203,159 @@ def forward(self,
     return output
 
 
+class TransformerEncoder(nn.Module):
+  r"""TransformerEncoder is a stack of N encoder layers. Users can build the
+  BERT(https://arxiv.org/abs/1810.04805) model with corresponding parameters.
+
+  Args:
+      encoder_layer: an instance of the TransformerEncoderLayer() class.
+      num_layers: the number of sub-encoder-layers in the encoder.
+      norm: the layer normalization component (optional).
+      enable_nested_tensor: if True, input will automatically convert to
+        nested tensor (and convert back on output). This will improve
+        the overall performance of TransformerEncoder when padding
+        rate is high.
+
+  Examples::
+    >>> encoder_layer = nn.TransformerEncoderLayer(12, 8)
+    >>> transformer_encoder = nn.TransformerEncoder(encoder_layer, 6)
+    >>> src = torch.rand(10, 32, 512)
+    >>> out = transformer_encoder(src)
+  """
+  __constants__ = ['norm']
+
+  def __init__(self,
+               encoder_layer,
+               num_layers,
+               norm=None,
+               enable_nested_tensor=True,
+               mask_check=True):
+    super().__init__()
+    self.layers = nn.ModuleList(
+        [copy.deepcopy(encoder_layer) for _ in range(num_layers)])
+    self.num_layers = num_layers
+    self.norm = norm
+    self.enable_nested_tensor = enable_nested_tensor
+    self.mask_check = mask_check
+
+  def forward(self,
+              src: Tensor,
+              mask: Optional[Tensor] = None,
+              src_key_padding_mask: Optional[Tensor] = None) -> Tensor:
+    """Pass the input through the encoder layers in turn.
+
+    Args:
+        src: the sequence to the encoder (required).
+        mask: the mask for the src sequence (optional).
+        src_key_padding_mask: the mask for the src keys per batch (optional).
+
+    Shape:
+        see the docs in Transformer class.
+    """
+    if src_key_padding_mask is not None:
+      _skpm_dtype = src_key_padding_mask.dtype  # pylint: disable=invalid-name
+      if _skpm_dtype != torch.bool and not torch.is_floating_point(
+          src_key_padding_mask):
+        raise AssertionError(
+            'only bool and floating types of key_padding_mask are supported')
+    output = src
+    convert_to_nested = False
+    first_layer = self.layers[0]
+    src_key_padding_mask_for_layers = src_key_padding_mask
+    why_not_sparsity_fast_path = ''
```

Comment:
As mentioned in our meeting earlier, a lot of the code here can be simplified a lot since the conditions for using the fast path implementation will never be fulfilled in the workload anyway. But we can do this in a later clean-up PR.